http/src/main/java/io/micronaut/http/body/ByteBody.java

```diff
@@ -0,0 +1,212 @@
+ * <p>An individual {@link ByteBody} is <i>not</i> thread-safe: You may not call
+ * {@link #split()} concurrently from multiple threads for example. However, the new
+ * {@link ByteBody} returned from {@link #split()} is independent, so you may use it on a
+ * different thread as this one.
+ *
+ * @author Jonas Konrad
+ * @since 4.5.0
+ */
+@Experimental
+public interface ByteBody {
+    /**
+     * Equivalent to {@code split(SplitBackpressureMode.SLOWEST)}.
+     *
+     * @return The newly split body. Must be closed by the caller, unless a terminal operation is
+     * performed on it
+     */
+    @NonNull
+    default CloseableByteBody split() {
+        return split(SplitBackpressureMode.SLOWEST);
+    }
+
+    /**
+     * Create a new, independent {@link ByteBody} that contains the same data as this one.
+     *
+     * @param backpressureMode How to handle backpressure between the old and new body. See
+     *                         {@link SplitBackpressureMode} documentation
+     * @return The newly split body. Must be closed by the caller, unless a terminal operation is
+     * performed on it
+     */
+    @NonNull
+    CloseableByteBody split(@NonNull SplitBackpressureMode backpressureMode);
+
+    /**
+     * Signal that the upstream may discard any remaining body data. Only if all consumers of the
+     * body allow discarding will the body be discarded, otherwise it will still be sent to all
+     * consumers. It is an optional operation.
+     * <p>Discarding may be implemented e.g. by closing the input side of an HTTP/2 stream.
+     * <p>This method must be called before any primary operation.
+     *
+     * @return This body
+     */
+    @Contract("-> this")
+    @NonNull
+    default ByteBody allowDiscard() {
+        return this;
+    }
+
+    /**
+     * Get the expected length of this body, if known (either from {@code Content-Length} or from
+     * previous buffering). The actual length will never exceed this value, though it may sometimes
+     * be lower if there is a connection error.
+     * <p>This value may go from {@link OptionalLong#empty()} to a known value over the lifetime of
+     * this body.
+     * <p>This is <i>not</i> a primary operation and does not modify this {@link ByteBody}.
+     *
+     * @return The expected length of this body
+     */
+    @NonNull
+    OptionalLong expectedLength();
+
+    /**
+     * Get this body as an {@link InputStream}.
+     * <p>This is a primary operation. After this operation, no other primary operation or
+     * {@link #split()} may be done.
+     *
+     * @return The streamed bytes
+     */
+    @NonNull
+    InputStream toInputStream();
+
+    /**
+     * Get this body as a reactive stream of byte arrays.
+     * <p>This is a primary operation. After this operation, no other primary operation or
+     * {@link #split()} may be done.
+     *
+     * @return The streamed bytes
+     */
+    @NonNull
+    Publisher<byte[]> toByteArrayPublisher();
```

Comment:
It is intentional that ImmediateByteBody extends the streaming case, so that users write their code for the streaming case but we can give optimized implementations where possible.